### DISCUSSION TITLE
Add a standard wallet interface

### DISCUSSION THREAD

**Discussion by piyalbasu** [stellar/stellar-protocol#1467] [Stellar Ecosystem Proposals]
*Background*
Many browser wallets expose an API so dapps can interact with a user's wallet. This allows a dapp to ask a wallet to do things like transmit a public key or sign an XDR.

*Problem*
Right now wallets all implement this API differently, which can make building a dapp or an SDK difficult; a dev must account for all these differences and manually integrate different wallet API's in order to support as many wallets as possible. This will conceivably become more complex as more wallets enter the ecosystem.

*Proposal*
The ecosystem agrees to a standard wallet API interface to make wallet integration easier. If all wallets agree to a minimum standard interface, a dapp or SDK developer could reliably infer what methods will be supported and what the function signature will be.

Example API interface:
```
{
  getPublicKey: () => Promise<publicKey as string>;
  signTransaction: (xdr: string) => Promise<signedXdr as string>;
  signAuthEntry: (authEntryXdr: string) => Promise<signedAuthEntry as string>
}
```

Relevant discord conversation: https://discord.com/channels/897514728459468821/1019346446014759013/1206810627377467413

**Comment by Shaptic** [comment]
It seems like the Discord discussion is pretty dated, so I'll offer some thoughts on this here, also to get this discussion going again.

I like the proposed interface, but I'd like to suggest an additional generic signing method that could actually encapsulate the other two and also be used for arbitrary signing purposes:

```typescript
interface Wallet {
  sign: (arbitraryB64: string) => Promise<string>;
}
```

(Internally, an implementation could have something like `signTransaction = (xdr: string) => { /* verify xdr is a Transaction, then just */ return this.sign(xdr); }`.)

-----

Also, should we discuss the trade-off between allowing encoded XDR passed in vs. the actual XDR object? I think encoded is fine, but we need to be sure to communicate to the implementers that they should NOT trust the input string to be well-formed or even necessarily represent an `xdr.Transaction`, for example. In other words, they should be very robust to failure and skeptical of the input. This wouldn't apply to the `sign` method, of course, besides maybe validating that it's base64?

**Reply by piyalbasu** [reply]
I like the idea of a generic signer. 

We have run into some issues passing the actual XDR object using `window.postMessage`. It appears postMessage strips the custom prototypes and will just pass an object with plain key/values. We've found it's more reliable to pass the xdr string and decode/encode at the destination

**Reply by Shaptic** [reply]
Ah, that's a great reason! Keeping it encoded is fine, then, we'll just have to remember the "tell people to not trust the input string" part.

**Reply by leighmcculloch** [reply]
Another motivation to stick with XDR is that to rely on the xdr.Transaction object is to make the interface only work with applications that use the js-stellar-sdk/base lib, which an application may not be using.

**Reply by leighmcculloch** [reply]
A generic signing function presents a couple challenges:

1. It'd need a parameter to discriminate which XDR type was being passed in. The discriminant is needed because XDR encoded does not self-identify. An XDR binary stores no data to tell you what type it is. In the case of TransactionEnvelope and AuthEntry you probably can tell by trying to decode one, failing, and trying to decode the other one. But if that works reliably, it is luck, as there's nothing in XDR that guarantees that distinction is possible.
   
   ```typescript
   interface Wallet {
      sign: (type: string, arbitraryB64: string) => Promise<string>;
   }
   ```
   
   When there are two separate functions the functions act as a discriminant, so really the generic signing function with the extra parameter is just moving the discriminant, not eliminating it.
   
   I think the generic function is more ergonomic, but I think it needs the discriminant still.

2. It makes it easier for a wallet developer to do an unsafe thing, because a wallet developer could look at that interface and think that they should support signing any arbitrary data. To do so would probably allow an application to get the user to sign things they shouldn't. So +100 to @Shaptic's comments about not trusting the input, it would be important for the wallet to be strict at only signing data that does decode to a tx or auth entry.

**Reply by piyalbasu** [reply]
@leighmcculloch That's great context, thanks. Without investigating, I had assumed that maybe there was a way to identify a xdr type from the xdr itself. Also very interesting that trying to decode as different types until success wouldn't reliably work either

We can still create a generic `sign` with the discriminant param. It is a little more dev experience friendly to just be able to change a param in the code rather than having to call a different method

**Reply by johansten** [reply]
> I like the idea of a generic signer.
> 
> We have run into some issues passing the actual XDR object using `window.postMessage`. It appears postMessage strips the custom prototypes and will just pass an object with plain key/values. We've found it's more reliable to pass the xdr string and decode/encode at the destination

You certainly should pass strings in window.postMessage/chrome.runtime.sendMessage if you want things to arrive like you sent them, but the *API* could still have met users where users are. Users are dealing with stellar.Transactions, not xdrs.

**Comment by earrietadev** [comment]
If we are adding extra methods, I would like to include an extra one:

- `signMessage` in the same way Albedo does it (or at least something similar).

I also prefer separated functions instead of a single generic function because it's easier for a wallet implementing something unique that they only have (like albedo with its messages signing features) and dapp developers can easily decide if they want to support that or not by just adding that extra function.

But that being said the more "features" we add (sighAuthEntry included) the more we will reduce the support for non stellar native wallets like those using wallet connect (Lobstr), hardware wallets, airgapped wallets, etc so at the end of the day devs will still need to handle different interfaces no matter what

**Comment by orbitlens** [comment]
I have mixed feelings about such an interface. It's simple and unified, and that's great. On the other hand, I really don't like the idea of returning plain strings from the method calls. 

Can we change the interface to return objects (with the predefined set of fields) instead of strings? For example, 

```ts
{
  getPublicKey: () => Promise<{publicKey: string}>,
  signTransaction: (xdr: string) => Promise<{signedXdr: string}>,
  signAuthEntry: (authEntryXdr: string) => Promise<{signedAuthEntry: string}>,
  signMessage: (message: string) => Promise<{signedMessage: string}>
}
```

Motivation:
- More flexible -- wallet developers can add additional fields to the response object without breaking the interface.
- It's future-proof and extendable -- if needed, the returned object format can can be extended in the future by adding additional fields, or removing obsolete fields.
- Backwards-compatible -- we'll be able to extend Albedo client lib bindings to support this standard without breaking other developers' code that relies on a specific format that we currently return.

**Comment by orbitlens** [comment]
Aside from the successful result standardization, we also have to define the standardized error output format.
External apps that integrate the wallet interface need to react on possible errors. Returning a generic error is bad idea, because every wallet will have a different error descriptions attached to it. The average app reaction on an error is to show the notification with error text. But depending on the connected wallet, the error text may be quite different. Ranging from "Failed" to "fetch() request error at line 13:53". This is non-informative for end-users and doesn't allow the client app to employ different logic for handling different types of errors (for example, in case of the Horizon error the client app might decide to resubmit the request again).

Albedo client lib defines [several standard errors](https://github.com/stellar-expert/albedo/blob/master/intent/src/intent-errors.js) for different scenarios. I'm not advocating for sticking with our vendor format, but it would be nice to have a few basic standard errors defined in a minimalistic format, e.g.

```ts
{
  message: string,    // general description message returned to the client app
  code: number,       // unique error code
  ext: Array<string>  // optional extended details
}
```

Proposed standard errors:
```js
{
  unhandled: { // internal wallet error, likely caused by the wallet logic itself
    message: 'Unhandled error occurred.',
    code: -1
  },
  external: { // external service (Horizon, RPC, etc.) returned an error 
    message: 'External error occurred.',
    code: -2,
    ext: ['Operation 2: Invalid "amount" value.', "Operation 5: Asset issuer is required."] // malformed tx error example
  },
  invalid: { // client app request is invalid (wrong parameters, invalid transaction XDR, etc.)
    message: 'Request is invalid.',
    code: -3,
    ext: ['Invalid transaction XDR'] // example error
  },
  rejected: { // user rejected the request, client app should not try to retry the request
    message: 'Rejected by the user.',
    code: -4
  }
}
```

**Comment by orbitlens** [comment]
One more thing to consider. Multi-account wallets in some cases need an option to identify the specific account to utilize for the request. 
For example, I have 10 different accounts in my wallet. When the client app requests the public key, I can save a "session" for this app, and imply that e.g. `signTransaction()` request relates to that exact account. 

However, if the user interacted with the app from two different accounts (for example, traded with DEX from two different accounts), I have a problem. The obvious solution is to show the list (or dropdown) of all wallet accounts with pre-selected the most recent account. Yet allowing an app to request signature for a specific account makes the UX much more streamlined on the wallet side, decreasing the possibility of a human error and saving a few clicks.

How about adding a second optional parameter `publicKey` into `signTransaction()`, `signAuthEntry()`, `signMessage()` methods to simplify interaction for multi-account wallets?

**Reply by earrietadev** [reply]
I agree with this and maybe instead of adding a second parameter we send an object as first parameter right away, it gives more flexibility if in the future we want to add something else

**Comment by orbitlens** [comment]
Lastly, kudos to @earrietadev for maintaining the [StellarWalletsKit](https://github.com/Creit-Tech/Stellar-Wallets-Kit) lib that allows client apps to avoid complexity of dealing with different wallets by abstracting the interface.

**Comment by leighmcculloch** [comment]
One thing I notice about the proposed API is that it doesn't expose any concept of a network, or a network passphrase. If a web app knew that it was only deployed to testnet or pubnet, how would it communicate the supported network, or see what network the wallet supported?

**Reply by earrietadev** [reply]
All 4 extension wallets (Albedo, Freighter, Rabet and xBull) already support sending the network as a parameter, that could be an option.

For example in [my kit](https://github.com/Creit-Tech/Stellar-Wallets-Kit/blob/50544bc1e6c1cb4098ce97186d97e3564977a46c/src/types.ts#L80) I have the function:
```ts
signTx(params: { xdr: string; publicKeys: string[]; network: WalletNetwork }): Promise<{ result: string }>;
```

This way the dapp can always tell the wallet the exact network it needs to use

**Reply by piyalbasu** [reply]
The interface above was definitely a bit simplified to kickoff the discussion. IMO, an actual interface would look more like:
```
{
  getPublicKey: () => /* return value up for debate */,
  signTransaction(xdr: string, opts?: { network?: string, networkPassphrase?: string, accountToSign?: string }) =>  /* return value up for debate */,
  signAuthEntry(xdr: string, opts?: { network?: string, networkPassphrase?: string, accountToSign?: string }) => /* return value up for debate */,
  getNetwork() => {"network": string, "networkUrl": string, "networkPassphrase": string, "sorobanRpcUrl": string} ,
}
```

**Comment by namankumar** [comment]
We could draw some inspiration from Ethereum as well. 

[EIP-1193](https://eips.ethereum.org/EIPS/eip-1193) was introduced to standardize a dapp connecting with different wallets from different providers - which is something I believe this proposal is trying to achieve as well

[EIP-6963](https://eips.ethereum.org/EIPS/eip-6963) was introduced to let a dapp discover when a browser wallet was installed, which becomes important when a user has multiple wallets installed

https://medium.com/walletconnect/goodbye-browser-extension-wallet-wars-eip-6963-is-now-approved-3fe0bc7476c8

**Comment by janewang** [comment]
@piyalbasu I'd like to suggest some interface additions. It could be for next iteration of this SEP.

```
connect()
disconnect()
accountsChanged() - When the user changes to another account 
sendTransaction(xdr: string) - This is the case where dapp relies on the wallet to send the transaction onward to rpc, instead of letting the dapp to interact with an RPC (on mainnet a lightweight dapp has to register separately to an RPC, instead of relying on a wallet to send the tx)
```

**Reply by orbitlens** [reply]
Methods like `connect()`, `disconnect()`, and `accountsChanged()` can be easily implemented on the client app side, with greater flexibility.

Regarding `sendTransaction()`, we can probably introduce additional `submit: boolean` parameter in the `signTransaction()` method. That's what we do currently in Albedo.

**Reply by johansten** [reply]
As a way of migrating to better supporting multiple accounts, which is better handled in the client, I suggest adding a getAccounts() that returns a list of all accounts that the user has authorized the wallet to share. By default, that could be one account, and everything works as before.

`accountsChanged()` isn't a bad idea, although more generally maybe we should have an `on()` for future safety.
Even for single accounts this is a good idea: You go to a website today, with "insert name" wallet adapter, and "insert name" wallet.
When change your active account in the wallet, the adapter has no way of knowing, so you have to reload the web page to get your updated state. That's not really a smooth UX, and we should be able to do better.

**Reply by johansten** [reply]
Oh, and the reason it's better dealt with in the client, is that as a user you want to minimize mental context switches.
There are actual context switches going on also, like starting up your wallet UI, which in some cases takes seconds.

**Comment by piyalbasu** [comment]
I've incorporated changes we've agreed to here as well as those I've collected on Discord on this PR: https://github.com/stellar/stellar-protocol/pull/1478

**Comment by earrietadev** [comment]
Just wanted bring something else to the table that it might be good to include: a way to import an asset.

Since we are promoting the usage of assets list, something I have used a lot when using EVM wallets is clicking on a protocol's site and add the asset to my wallet (basically it adds the asset contract ID). This is good for users that might be using a protocol that is not that popular yet or for custom assets that will probably be outside of the lists (like pool shares)

**Comment by janewang** [comment]
I'd like to suggest an `importAccountByPrivateKey` method to be implemented by all wallets. 

Occasionally wallets shutdown for various reasons (no more run way, key member departure, so on and so forth). Users could migrate off the wallet by using seedphrase, but in all instances, this is too challenging for _all_ the users. If wallets consistently have the ability to import account by private key, the wallet about to shutdown can push a feature to export keys and import to other wallets. It makes migration off a wallet less choatic. In general be able to import by key is useful from dev perspective too.

**Reply by willemneal** [reply]
Usually wallets use seed phrases. This helps users create multiple wallets and is much easier to export/import.

**Reply by piyalbasu** [reply]
This is actually something some others in the ecosystem have requested Freighter implement. We've decided not to implement this as we have felt we want to discourage users from copy/pasting private keys. 

cc: @brunomuler

**Reply by earrietadev** [reply]
I think there is value in making an standard for migrating form one wallet to another one, in xBull I already allow that by exporting an encrypted file, later I can import this to any xBull (mobile, web and extension) when I need to. If all wallets accept something similar I think we could make it easy for users to migrate if they need to and it can be done without them needing to copy/paste keys nor phrases

**Reply by janewang** [reply]
Yes, the key should be encrypted using a password during transfer. Alternatively, we can consider sweep an account. The point is aligning on something consistent such that it makes it easier to migrate users.

**Reply by piyalbasu** [reply]
I like the idea of an encrypted account migration. I think that it may be a bit out of scope for this standard, which focuses on exposing an API - maybe we should consider this for a different standard as I do think this is definitely important functionality to standardize!

I'd mostly be concerned about exposing this method to dapps, which could lead to scams. I'd rather this functionality be contained to the wallet where only the wallet owner has access to it

**Comment by mtwtfss** [comment]
@earrietadev and @orbitlens we're planning to finalize this proposed standard during the Stellar Protocol Meeting next week on Thursday, June 6th at 1:00 PM PDT. We'd love your support and participation in the meeting.

**Reply by johansten** [reply]
What happened? :)

**Reply by earrietadev** [reply]
@johansten https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0043.md

**Reply by johansten** [reply]
Yeah, I've seen that. But there's nothing about how to actually use it, in a browser. How do I get exposed to this API? How do I inject my wallet into this?

**Comment by namankumar** [comment]
The corresponding [proposal](https://github.com/stellar/stellar-protocol/blob/b89cf8b3e70367bed475f2bcbbfc1fec99df5227/ecosystem/sep-0043.md) was put to vote on 6/6/24. The votes represent ecosystem sentiment and can be used by the author to determine next steps 

The voting card is on testnet deployment of Soroban Governor ([link]( https://testnet.governance.script3.io/CC7XSM5U5Z7UGC3BMN5IHDH4R7DTRE4NFE46SKNSFN2SIMQZ6AVWTGEW/proposal/?id=6))

Here is the snapshot of the final count:

![Screenshot 2024-06-07 at 3 41 59 PM](https://github.com/stellar/stellar-protocol/assets/330364/137d9351-60cf-43b1-a8b4-bc39f87d50f0)

![Screenshot 2024-06-07 at 3 42 23 PM](https://github.com/stellar/stellar-protocol/assets/330364/72397d40-a504-443a-9608-cf1b87e31288)

**Comment by namankumar** [comment]
I'm curious which wallets are planning to implement or have already implemented this standard? @piyalbasu @earrietadev @orbitlens

**Reply by earrietadev** [reply]
This interface is included in the Stellar Wallets Kit, xBull already supports part of the interface but not all of the methods yet

**Comment by orbitlens** [comment]
Do we really need `signAuthEntry` method? What's the case for it? 

Someone mentioned earlier that it can be useful for multisig. However, to my mind, `signTransaction` must handle all multisig aggregations instead. This way developers don't need to think which method they need to utilize for a particular transaction and how the multisig or multi-authorization should be handled.

**Reply by piyalbasu** [reply]
`signAuthEntry` is mostly used for atomic swaps on Soroban. A user needs to sign just for the authentication of their part of the swap, not the whole transaction. We had talked internally about creating a generic signing method where a caller doesn't need to choose the type of payload to sign, but it seems as though this may not be possible: https://github.com/stellar/stellar-protocol/discussions/1467#discussioncomment-8944725

Because of that, we opted for a separate signAuthEntry method

**Comment by orbitlens** [comment]
How `getAddress()` method should behave if the wallet contains more than one account and a user previously interacted with this site using several different addresses? Should it be the last utilized address?

Do we want to make sure that a user _**really owns**_ the account he authenticating with? I mean, someone can easily clone the open-source wallet repo, and slightly change the code to impersonate some other account owner, returning a public key from the account they doesn't own (e.g., one of SDF's accounts). We have two options: either sign and return some challenge in this method, similar to the `signMessage()` method, or simply mention in SEP that `getAddress()` doesn't guarantee the account ownership and if developers need to authenticate an account owner, they should use `signMessage()` method for this purpose.

**Reply by earrietadev** [reply]
IMO `getAddress` should just return the selected account on the wallet and about the impersonation part, the app should be aware of that and be the one that takes care of making sure an account is actually owned by the user (for example using SEP-10 if the app needs to show something other accounts should't see)

**Reply by johansten** [reply]
Why does a wallet need an active account (and/or active network)? Maybe even more important, why should you select the active account in the wallet, and not in the dapp, which is where the user is otherwise focusing their attention?

**Reply by earrietadev** [reply]
@johansten because wallets are standalone apps that can be used without the need of an external app (some of them don't even connect with external apps), and so they have their own "active account" logic (and with wallets where you can have multiple networks then it's even more important). On the other hand a dapp can also just tell the wallet what account to use when requesting a signature.

So for both questions `Why does a wallet need an active account` and `why should you select the active account in the wallet, and not in the dapp, which is where the user is otherwise focusing their attention` the answer will be: It doesn't matter because both cases are supported by wallets.

**Reply by johansten** [reply]
I've spent quite some time developing wallets, so I know how they work, and that's why I'm asking. There's a lot of bad UX going around in this space, and active account/active network has to be one of the worst. If you've ever used Ethereum for multiple networks, e.g., you've surely come across the situation where you have your metamask, and you have two dapps open in two tabs. The dapps are on different networks though, so *even though the wallet itself can manage multiple networks at the same time*, web3js only lets you have one active account/network globally, so the dapp that has the non-active network is just in limbo. Same thing, but to a lesser degree, when you have a wallet with multiple accounts; you have to go to the dapp, and click a button to switch accounts/network before it's usable again. So I'm very much opposed to these bad UX patterns, and the APIs that impose them. I think @orbitlens has a similar comment in this thread, where he thinks the network *for a tab* should be selected in the dapp, and not in the wallet.

Instead of getAccount/getNetwork, how about a getAccounts/getNetworks, *similar* to what Metamask et al. does with connecting accounts to sites, where you as a wallet user can select which accounts you want a dapp to get exposure to, and then the dapp can pick which ever it wants, whenever it wants, without affecting the internal state of the wallet. I think that's the key here. State. You shouldn't have one global state, each dapp should have its own. The wallet UI can have one account open, and one dapp has another one, a second dapp can have a different one, and when you get a signing request, you just display which account/network it's for.

**Reply by earrietadev** [reply]
@johansten The examples you're using are not relevant with Stellar: 
- You don't have multiple networks like in the EVM space where you are dealing with a ton of layer 1s and 2s, here you are either on testnet or mainnet, that's all and most users don't even care about testnet.
- The wallets on stellar don't care if your dapp is asking for another account/network that is not the current "active" account on the wallet, the wallet will just use whatever the dapp says and sign with it without caring about anything else.

> The wallet UI can have one account open, and one dapp has another one, a second dapp can have a different one, and when you get a signing request, you just display which account/network it's for.

That's exactly how it works with most of the Stellar wallets

**Reply by johansten** [reply]
It was an example of how API design can have unintended downstream consequences on UX.
Why then have a getaddress() instead of getaddresses() in the proposed API?
"IMO getAddress should just return the selected account on the wallet"
That's how we get into the bad UX territory. Now every dapp is dependent on wallet UI view, and the user switching between the dapp and the wallet to do state changes.

**Reply by orbitlens** [reply]
@earrietadev Just for context, @johansten is an OG and the creator of a one of the first Stellar wallets.

**Reply by earrietadev** [reply]
@orbitlens I'm aware of it

@johansten 
> That's how we get into the bad UX territory. Now every dapp is dependent on wallet UI view, and the user switching between the dapp and the wallet to do state changes.

The proposed method was "getaddress" not "getaddresses"... There is always the possibility to just extend the standard if is that big of a deal. But to be honest even if "getaddresses" were proposed I'm not sure I will be adding it to xBull, why? because I have 741 stellar accounts across 37 wallets on different networks in xBull, most sites won't even be able to handle them. Heck most apps don't even keep one single account across sessions in the app so now imagine asking devs to even show a proper UI to select one of your accounts

My case is of course an extreme case because not everybody is that active on the network, but my point is that not all wallets handle all accounts in the same way because in my case xBull allows you to set names to "wallets" and "accounts" inside of it to better organize everything, asking dapps to do the same so they are compatible with xBull "getaddresses" method is not going to be easy because most of them will just pass (like they already do by not caring about accounts management)

**Reply by johansten** [reply]
If we don't do a wallet API that supports multiple accounts now, will the dapps *ever* adapt?
"The best time to plant a tree...", etc, etc..

That was the only reason I did Stargazer back the in the days. The 2015 protocol was so ahead of its time it was insane, and the other wallets at the time, primarily Centaurus and Lobstr, hardly even scratched the surface in terms of what they could do. I think it took them years before the even did non-native assets. And from the start, I did non-native assets, I did multiple accounts, I did transactions with multiple source accounts, I did source accounts with multiple signers, the lot. Because the protocol deserved it.

**Comment by orbitlens** [comment]
We probably need to standardize the implementation of the `signMessage()` method.

Our algorithm: 
```js
function signMessage(message, keypair) {
  const publicKey = keypair.publicKey()
  const signedMessage = keypair.publicKey()+':'+message //protection from spoofing
  const messageHash = sha256(signedMessage)
  const signature = keypair.sign(messageHash).toString('hex')
  return {publicKey, signature, signedMessage}
}
```

From my point of view, the `networkPassphrase` argument is irrelevant here since it's a general-purpose method, not tied to a specific Stellar network. If developers want to make sure that the user is signing message for a specific network, they can include network identifier into the `message` argument.

**Reply by overcat** [reply]
**Update:** See https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0053.md

We should ensure that any string (Chinese, Japanese, etc.) can be properly processed. Also, I think using `Stellar Signed Message:\n'` as a prefix could be another option, as this format is widely adopted. Here's a possible implementation:


```js
function signMessage(message, keypair) {
    const signedMessage = 'Stellar Signed Message:\n' + message;
    const messageBuffer = Buffer.from(signedMessage, 'utf8');
    
    const messageHash = sha256(messageBuffer);
    
    const signature = keypair.sign(messageHash);
    const publicKey = keypair.publicKey();
    
    return {
        publicKey,
        signature,
        signedMessage
    };
}
```

**Comment by orbitlens** [comment]
Regarding the `getNetwork()` method. We don't support this method currently because it can potentially cause weird situations when a user temporarily switched to the Testnet and after that can't sign a transaction in a dapp interface that works with the Pubnet.

As I see it, the network selection during the transaction signing should be controlled by the dapp UI. A user should see the requested network in the wallet interface during the request authorization, but shouldn't be able to change it (again, to avoid weird glitches). The network, selected by the user in the wallet interface prior to the transaction signing request must not affect the request received from the dapp. If the network was not specified in the request, the wallet should assume that the dapp uses Stellar Pubnet by default.

**Reply by piyalbasu** [reply]
Do you think there's a security implication here? For example, a dapp could claim that their transaction is only meant for testnet, but then actually configure for mainnet. The user would have the opportunity to see the discrepancy in the wallet's UI, but they might miss it if they're not careful

**Comment by johansten** [comment]
FWIW, I started defining a version of Wallet Standard, https://wallet-standard.github.io/wallet-standard/, for Stellar.

One more standard, you say?

\<_insert obligatory xkcd strip_\>

The naming is a bit presumptuous, but as a spec, "Wallet Standard" simply slaps.

It's the most sane browser extension wallet framework I've seen.

I think emphasizing *framework* here is important, because this is an end-to-end solution. It's not just an interface to implement, it's the supporting framework to use your interface within.

It covers what you need to do as a wallet developer to implement your interfaces, and how to register them.

```
interface Wallet {
    readonly version: WalletVersion;
    readonly name: string;
    readonly icon: WalletIcon;
    readonly chains: IdentifierArray;
    readonly features: IdentifierRecord<unknown>;
    readonly accounts: readonly WalletAccount[];
}

interface WalletAccount {
    readonly address: string;
    readonly publicKey: ReadonlyUint8Array;
    readonly chains: IdentifierArray;
    readonly features: IdentifierArray;
    readonly label?: string;
    readonly icon?: WalletIcon;
}

....

import { registerWallet } from '@wallet-standard/core';

registerWallet(wallet);
```
It covers what you need to do as a dapp developer to discover registered interfaces, and how to access them.

```
import { getWallets } from '@walllet-standard/base';
import { StellarWallet, ... } from '@majestic-suitcase/core'; 

const { get, on } = getWallets();
const wallets = get();

... filter for wallets that exposed stellar features, and stellar chains ...

const wallet: StellarWallet = wallets[0] ;

```

Nothing about signing transactions?

Well, Wallet Standard is a framework, right? There's what's called "features", with which a wallet can expose functions, and information about those functions, e.g. versioning.

There's a few standard features that wallets have to implement, `standard:connect`, `standard:disconnect`, `standard:events`, for, you guessed it, connecting/disconnecting a wallet, and managing event handlers.

```
const wallet = {
    ...
    features: {
        ...
        'standard:connect': {
            version: '1.0.0',
            connect: this.#connect,
        },
        ...
    }
    ...
};

const accounts = await wallet.features['standard:connect'].connect();
```

Other than that, it's a free-for-all.

With Solana defining this, they obviously had their perspective on things when they designed the interfaces for Solana wallets, and a lot of them make a lot of sense in general.

They rightfully take the stance that the wallet state has nothing to do with dapp state(s). Thus, when a user connects a wallet to a website, they decide which account(s) to expose to the dapp, and it's then up to the *website* to keep track of which account it's currently using, and pass that along with any call to the wallet. There is no global, currently active account. There is no global, currently active network. Changing network in one dapp, shouldn't affect either the wallet or any other dapp.

Now, I'm not sure if I agree with their idea of using arrays for inputs/output, i.e., always allow the user so sign an arbitrary amount of things in one call, or with the idea of returning fully signed transactions instead of just a signature.

But yeah, you're free to go your own way. You can, e.g. look at Aptos, who took everything good about the Solana interface, and threw it out the window.

```
... the dapp requests user to sign a transaction, using a provided account

const res = await wallet.features['stellar:signTransaction'].signTransaction({
    account,
    chain: 'stellar:public',
    transaction
})
```

# Code

https://github.com/johansten/majestic-suitcase/
-   'core' has the wallet interfaces
-   'extension' is a chrome extension wallet that implement them
-   'example' is a webpage that connects to a Standard Wallet and signs things
-   'react' is a couple of React components that simplify usage for a dapp

# Things I've learned

- Wallet Standard really, really slaps. If you have an existing Standard Wallet for blockchain X, all you need to do is implement a `StellarWalletAccount`, add its features to your implementation of the `Wallet` interface, and you're set. The delta is so ridiculously tiny.
- The hard part of being a developer these days is in configuring your tools.

# Going forward

- I should probably be passing `stellar.Transaction`s to `signTransaction`, that way a user don't have to pass the network parameter twice, just when they build it.
- `useWalletContext()` should probably have helper functions to abstract away the way you use the exposed features, plus functionality for verifying the different kinds of signatures we have.

**Reply by johansten** [reply]
Example web at https://johansten.github.io/majestic-suitcase/
Get the extension from https://johansten.github.io/majestic-suitcase/extension.tar.gz
Chrome only, AFAIK.
